Missing Post Type supports in Options Modal

imath opened the pull request:
- Featured Image is not only a Theme thing, it is possible for a custom post type to declare itself without supporting the `thumbnail` feature.
- A custom post type can also register itself withour supporting `comments` and/or `custom-fields`.

As a result the corresponding Document panels in the editor sidebar are not displayed. The Options Modal should also avoid displaying the panel checkboxes if the custom post type does not support these 3 features

## Description
To fix this issue, i've wrapped featured image, comments and custom fields checkboxes into `<PostTypeSupportCheck/>` components

## How has this been tested?
I've tested the PR with the `wp_block` post type which only supports the `title` and `editor` features and check everything was still behaving the right way with the `post` post type.

## Screenshots

### Without the PR applied:

![without](https://user-images.githubusercontent.com/1834524/49177892-a8cff600-f34e-11e8-8435-8890dfd03a4b.png)

Checkboxes for featured image, comments, and custom fields are displayed although they shouldn't

### With the PR applied:

![with](https://user-images.githubusercontent.com/1834524/49177978-d9179480-f34e-11e8-8221-f91c9edab4ef.png)

The checkboxes for these 3 unsupported features are no more displayed.

## Types of changes
Bug fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

imath committed:
Missing Post Type supports in Options Modal
Featured Image is not only a Theme thing, it is possible for a custom post type to declare itself without supporting the `thumbnail` feature.
A custom post type can also register itself withour supporting `comments` and/or `custom-fields`.
As a result the corresponding Document panels in the editor sidebar are not displayed. The Options Modal should also avoid displaying the panel checkboxes if the custom post type does not support these 3 features

imath committed:
Do not use single quotes in attributes

ajitbohra reviewed (approved):
LGTM :+1:

gziolo commented:
Should we also wrap the permalink section with check? It Was added in #12132.

imath committed:
Display the Post permalink panel option only if the permalink is editable

imath committed:
Update the Options Modal snapshots

imath commented:
@gziolo Absolutely. Thanks for mentioning it. I've just updated the PR accordingly.
+ I've updated the Options Modal snapshots that were failing.

noisysocks commented:
It would be nice if we didn't display the section title when there are no options underneath it. This came up in https://github.com/WordPress/gutenberg/pull/11802#issuecomment-438481134 as well.

noisysocks commented on packages/edit-post/src/components/options-modal/meta-boxes-section.js:
```diff
@@ -25,7 +26,11 @@ export function MetaBoxesSection( { areCustomFieldsRegistered, metaBoxes, ...sec
 		<Section { ...sectionProps }>
-			{ areCustomFieldsRegistered && <EnableCustomFieldsOption label={ __( 'Custom Fields' ) } /> }
+			{ areCustomFieldsRegistered && (
+				<PostTypeSupportCheck supportKeys="custom-fields">
```
This check shouldn't be necessary because `getEditorSettings().enableCustomFields` should be `undefined` when the current post type doesn't support custom fields. 

It looks like this works correctly in WordPress 5.0 RC2 but not in WordPress 4.9 + Gutenberg 4.6.1.

I think that we should remove this check and make `enableCustomFields` work the same as it does in core.

Alternatively, we could remove this check and then simply wait for the Gutenberg plugin to require WordPress 5.0 as its minimum supported version.

imath replied:
I agree about
> I think that we should remove this check and make enableCustomFields work the same as it does in core.

That's what I did in https://github.com/WordPress/gutenberg/pull/12394/commits/8a3b3f04164fb3968fc557afd03f1210a92b7fc0

gziolo commented:
> It would be nice if we didn't display the section title when there are no options underneath it. This came up in [#11802 (comment)](https://github.com/WordPress/gutenberg/pull/11802#issuecomment-438481134) as well.

Yes, I opened #11923 to track it.

jorgefilipecosta commented on packages/edit-post/src/components/options-modal/index.js:
```diff
@@ -42,7 +42,9 @@ export function OptionsModal( { isModalActive, closeModal } ) {
 			</Section>
 			<Section title={ __( 'Document Panels' ) }>
-				<EnablePanelOption label={ __( 'Permalink' ) } panelName="post-link" />
+				{ isPermalinkEditable && (
```
Even if permalink is not editable we may show the Permalink panel. The conditions to show the permalink panel are complex e.g: we don't show it for new posts etc.
I think the ideal check for this case is ```<PostTypeSupportCheck supportKeys={ [ 'viewable' ] }>``. This way we don't show the option in post types where the permalink panel will never appear, but we show the option for post type where the permalink may appear even if it doe not appear currently e.g. because the post is new.

imath replied:
Thanks for your feedback, that's what I did in https://github.com/WordPress/gutenberg/pull/12394/commits/beda76c02447b1e0547720be4c1d695a0df46c5e

imath committed:
Make enableCustomFields work the same as it does in core.

As recommanded by @noisysocks : make sure the behavior about custom fields is consistent between WordPress 5.0 RC and the Gutenberg plugin

imath committed:
Remove the unused dependency to @wordpress/editor

It should fix the failing test.

imath committed:
Use the viewable post type property to check whether to show the permalink modal options checkbox

Apply @jorgefilipecosta advice.

noisysocks commented:
Let's tackle https://github.com/WordPress/gutenberg/issues/11923 in a seperate PR. I don't think it's urgent.

imath commented:
Fine with me @noisysocks, should I revert https://github.com/WordPress/gutenberg/pull/12394/commits/20d1669b45b3e59b5ae876e62d218c3687d7280f @gziolo ?

gziolo commented:
> Fine with me @noisysocks, should I revert 20d1669 @gziolo ?

Yes, I would put it in a follow-up PR so we could iterate on it 👍

imath commented on lib/client-assets.php:
```diff
@@ -1030,7 +1030,7 @@ function gutenberg_get_available_image_sizes() {
 	$is_demo = isset( $_GET['gutenberg-demo'] );
 
-	global $wp_scripts;
+	global $wp_scripts, $wp_meta_boxes;
```
Actually, the part I've added to disable custom fields is already in Core since branch 5.0. So it's more an update into Gutenberg to make sure dealing with custom fields is done the same way than in Core.

see https://core.trac.wordpress.org/browser/branches/5.0/src/wp-admin/edit-form-blocks.php#L364

gziolo replied:
Right, I see it now. Thanks for explanation 👍

imath commented on packages/edit-post/src/components/options-modal/index.js:
```diff
@@ -25,24 +25,25 @@ import MetaBoxesSection from './meta-boxes-section';
+	const hasPageAttributeSupport = get( postType, [ 'supports', 'page-attributes' ], false ) || hasTemplates;
+	let documentPanelsSection;
+
+	if ( isViewable || hasTaxonomies || hasFeaturedImageSupport || hasExcerptSupport || hasCommentsSupport || hasPageAttributeSupport ) {
```
😱

imath commented:
@gziolo I've just reverted the commit that was doing the huge check to try to solve #11923

gziolo reviewed (approved):
Looks good after all changes made 👍

gziolo commented:
@noisysocks - do you want to perform another round of testing before we proceed? It's blocked by your review at the moment :)

noisysocks commented:
@gziolo: Dismissed my stale review. Am doing a support rotation this week so best not wait for me 🙂

gziolo commented:
@imath - can you rebase with `master`? Travis is back to normal :)

imath commented:
@youknowriad sorry i’ve clicked on the resolve button before posting this comment.

There’s no need to backport as it is already there these lines simply moved down a few lines. You should see the removed lines highlighted in red into the changes.

imath commented:
And btw thanks a lot for the merge 😍
